In [1]:
import numpy as np
import pandas as pd
import sklearn
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

## **Loading dataset**

In [2]:
from google.colab import files
uploaded = files.upload()

Saving data.csv to data.csv


In [3]:
df = pd.read_csv("data.csv")
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

## **Exploring dataset**

In [4]:
df.shape

(569, 33)

In [5]:
# Checking datatypes
df.dtypes

id                           int64
diagnosis                   object
radius_mean                float64
texture_mean               float64
perimeter_mean             float64
area_mean                  float64
smoothness_mean            float64
compactness_mean           float64
concavity_mean             float64
concave points_mean        float64
symmetry_mean              float64
fractal_dimension_mean     float64
radius_se                  float64
texture_se                 float64
perimeter_se               float64
area_se                    float64
smoothness_se              float64
compactness_se             float64
concavity_se               float64
concave points_se          float64
symmetry_se                float64
fractal_dimension_se       float64
radius_worst               float64
texture_worst              float64
perimeter_worst            float64
area_worst                 float64
smoothness_worst           float64
compactness_worst          float64
concavity_worst            float64
concave points_worst       float64
symmetry_worst             float64
fractal_dimension_worst    float64
Unnamed: 32                float64
dtype: object

In [6]:
# Checking for nulls
nulls = df.isna().sum().reset_index()

In [ ]:
nulls

index    0
0                        id    0
1                 diagnosis    0
2               radius_mean    0
3              texture_mean    0
4            perimeter_mean    0
5                 area_mean    0
6           smoothness_mean    0
7          compactness_mean    0
8            concavity_mean    0
9       concave points_mean    0
10            symmetry_mean    0
11   fractal_dimension_mean    0
12                radius_se    0
13               texture_se    0
14             perimeter_se    0
15                  area_se    0
16            smoothness_se    0
17           compactness_se    0
18             concavity_se    0
19        concave points_se    0
20              symmetry_se    0
21     fractal_dimension_se    0
22             radius_worst    0
23            texture_worst    0
24          perimeter_worst    0
25               area_worst    0
26         smoothness_worst    0
27        compactness_worst    0
28          concavity_worst    0
29     concave points_worst    0
30           symmetry_worst    0
31  fractal_dimension_worst    0
32              Unnamed: 32  569

In [7]:
# Checking for duplicates
duplicates = df[df.duplicated()]

In [8]:
duplicates

Empty DataFrame
Columns: [id, diagnosis, radius_mean, texture_mean, perimeter_mean, area_mean, smoothness_mean, compactness_mean, concavity_mean, concave points_mean, symmetry_mean, fractal_dimension_mean, radius_se, texture_se, perimeter_se, area_se, smoothness_se, compactness_se, concavity_se, concave points_se, symmetry_se, fractal_dimension_se, radius_worst, texture_worst, perimeter_worst, area_worst, smoothness_worst, compactness_worst, concavity_worst, concave points_worst, symmetry_worst, fractal_dimension_worst, Unnamed: 32]
Index: []

[0 rows x 33 columns]

## **Feature Engineering**

In [9]:
# Dropping Unnamed column & id column
df = df.drop(['Unnamed: 32', 'id'], axis=1)

In [10]:
df.columns

Index(['diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean',
       'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst'],
      dtype='object')

In [11]:
# Checking unique values for diagnosis column
df['diagnosis'].unique()

array(['M', 'B'], dtype=object)

In [12]:
# Converting diagnosis column object to int
df['diagnosis'] = df['diagnosis'].map({'M' : 1, 'B' : 0})

In [13]:
df['diagnosis'].unique()

array([1, 0])

In [14]:
# Split data to X, y
X = df.drop(['diagnosis'], axis = 1)
y = df['diagnosis']

## **Feature Scaling**

In [15]:
# Scaling my X data
scaler = StandardScaler()

X_scaled = scaler.fit_transform(X)

## **Train & Test split**

In [16]:
# Split data to train and test
X_test, X_train, y_test, y_train = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

## **Model Training**
### **1. Logistic Regression**

In [19]:
# Training
log_model = LogisticRegression()
log_model.fit(X_train, y_train)
log_model_pred = log_model.predict(X_test)
log_accuracy = accuracy_score(y_test, log_model_pred)
print("Logistic Regression Accuracy:", log_accuracy)
print("Classification Report:\n", classification_report(y_test, log_model_pred))

Logistic Regression Accuracy: 0.9560439560439561
Classification Report:
               precision    recall  f1-score   support

           0       0.97      0.97      0.97       286
           1       0.94      0.94      0.94       169

    accuracy                           0.96       455
   macro avg       0.95      0.95      0.95       455
weighted avg       0.96      0.96      0.96       455



**Accuracy**: 95.6% — Overall, your model correctly predicted 95.6% of the cases.

**1. Precision:**
- Class 0 (negative): 97% — When the model predicts class 0, it is correct 97% of the time.
- Class 1 (positive): 94% — When the model predicts class 1, it is correct 94% of the time.

**2. Recall:**
- Class 0 (negative): 97% — Of all the actual class 0 cases, the model correctly identified 97% of them.
- Class 1 (positive): 94% — Of all the actual class 1 cases, the model correctly identified 94% of them.

**3. F1-Score:**
- Class 0: 0.97 — A balance between precision and recall for class 0.
- Class 1: 0.94 — A balance between precision and recall for class 1.

### **2. Random Forest**

In [20]:
rad_model = RandomForestClassifier()
rad_model.fit(X_train, y_train)
rad_model_pred = rad_model.predict(X_test)
rad_accuracy = accuracy_score(y_test, rad_model_pred)
print("Random Forest Accuracy:", rad_accuracy)
print("Classification Report:\n", classification_report(y_test,rad_model_pred))

Random Forest Accuracy: 0.9252747252747253
Classification Report:
               precision    recall  f1-score   support

           0       0.96      0.92      0.94       286
           1       0.88      0.93      0.90       169

    accuracy                           0.93       455
   macro avg       0.92      0.93      0.92       455
weighted avg       0.93      0.93      0.93       455



**Accuracy:** 92.5% — Overall, your model correctly predicted 92.5% of the cases.

**1. Precision:**
- Class 0 (negative): 96% — When the model predicts class 0, it is correct 96% of the time.
- Class 1 (positive): 88% — When the model predicts class 1, it is correct 88% of the time.

**2. Recall:**
- Class 0 (negative): 92% — Of all the actual class 0 cases, the model correctly identified 92% of them.
- Class 1 (positive): 93% — Of all the actual class 1 cases, the model correctly identified 93% of them.

**3. F1-Score:**
- Class 0: 0.94 — A balance between precision and recall for class 0.
- Class 1: 0.90 — A balance between precision and recall for class 1.

### **Saving model**

In [25]:
# Export model
import pickle

# Save files to Colab's local filesystem
with open('log_model.pkl', 'wb') as model_file:
    pickle.dump(log_model, model_file)

with open('scaler.pkl', 'wb') as scaler_file:
    pickle.dump(scaler, scaler_file)